archive/imo/imo2021_q1.lean

```diff
@@ -0,0 +1,299 @@
+/-
+Copyright (c) 2021 Mantas Bakšys. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mantas Bakšys
+-/
+
+import data.real.sqrt
+import tactic.interval_cases
+import tactic.linarith
+import tactic.norm_cast
+import tactic.norm_num
+import tactic.ring_exp
+
+
+/-!
+# IMO 2021 Q1
+Let n≥100 be an integer. Ivan writes the numbers n, n+1,..., 2n each on different cards.
+He then shuffles these n+1 cards, and divides them into two piles. Prove that at least one
+of the piles contains two cards such that the sum of their numbers is a perfect square.
+# Solution
+We show there exists a triplet a, b, c ∈ [n , 2n] with a < b < c and each of the sums (a + b),
+(b + c), (a + c) being a perfect square. Specifically, we consider the linear system of equations
+
+  a + b = (2 * l - 1) ^ 2
+  a + c = (2 * l) ^ 2
+  b + c = (2 * l + 1) ^ 2
+
+which can be solved to give
+
+  a = 2 * l * l - 4 * l
+  b = 2 * l * l + 1
+  c = 2 * l * l + 4 * l
+
+Therefore, it is enough to show that there exists a natural number l such that
+n ≤ 2 * l * l - 4 * l and 2 * l * l + 4 * l ≤ 2 * n for n ≥ 100.
+
+Then, by Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
+finishes the proof.
+-/
```

Comment:
Markdown (just in case we render this
```suggestion

/-!
# IMO 2021 Q1

Let `n≥100` be an integer. Ivan writes the numbers` n, n+1,..., 2n` each on different cards.
He then shuffles these `n+1` cards, and divides them into two piles. Prove that at least one
of the piles contains two cards such that the sum of their numbers is a perfect square.

# Solution

We show there exists a triplet `a, b, c ∈ [n , 2n]` with `a < b < c` and each of the sums `(a + b)`,
`(b + c)`, `(a + c)` being a perfect square. Specifically, we consider the linear system of equations

    a + b = (2 * l - 1) ^ 2
    a + c = (2 * l) ^ 2
    b + c = (2 * l + 1) ^ 2

which can be solved to give

    a = 2 * l * l - 4 * l
    b = 2 * l * l + 1
    c = 2 * l * l + 4 * l

Therefore, it is enough to show that there exists a natural number l such that
`n ≤ 2 * l * l - 4 * l` and `2 * l * l + 4 * l ≤ 2 * n` for `n ≥ 100`.

Then, by the Pigeonhole principle, at least two numbers in the triplet must lie in the same pile, which
finishes the proof.
-/
```

Reply:
It looks to me that some of the lines are over the character limit in your suggestion, does this not matter here?

Reply:
No, it just means that I didn't check them!

Reply:
Commited a version, where the limit is obeyed :)